### PR TITLE
Implement Pylint W0149

### PR DESCRIPTION
**Pull request by tjkuson**
## Summary

Implement the Pylint [W0149 rule](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/while-used.html) that checks for while loops. Related to #970.

Includes documentation.

## Test Plan

Added test files.

### CODE REVIEW

**Review by MichaReiser — COMMENTED** [review]
The code looks good to me but I'm conflicted on the rule itself. It seems over-aggressive to flag every while loop. Could we improve the rule to detect cases where the loop can be rewritten to a `for range` loop with reasonable certainty?

File: crates/ruff/src/checkers/ast/mod.rs [hunk 1]
```diff
@@ -1510,6 +1510,9 @@ where
                 if self.enabled(Rule::UselessElseOnLoop) {
                     pylint::rules::useless_else_on_loop(self, stmt, body, orelse);
                 }
+                if self.enabled(Rule::WhileLoop) {
+                    pylint::rules::while_loop(self, stmt);
+                }
             }
```
[the remark this answers is not in the document]
Comment (tjkuson) [reply crates/ruff/src/checkers/ast/mod.rs:1516]: Changed in [c786b8a](https://github.com/charliermarsh/ruff/pull/4652/commits/c786b8ad209af107d5bbccff2f9b38018649950e).

**Comment by tjkuson** [comment]
Is it worth creating a separate issue for this rule to discuss its merits? I assumed that the rule would be implemented as-is following #970. However, I recognise it might be a controversial rule.

For what it's worth, I value Pylint W0149 in my workflow as-is, and if the rule were to be adapted, I would prefer it to have limited exceptions rather than conservatively flagging `while` loops that can be cleanly rewritten to `for` loops.

**Comment by MichaReiser** [comment]
CC @charliermarsh

**Comment by charliermarsh** [comment]
I think I'd prefer not to implement this rule for now. It requires a Pylint extension (`pylint.extensions.while_used`), so it's not enabled by default, and I suspect most users would find it to be too aggressive to be on-by-default.

I'd be open to implementing it if we had the ability for users to selectively opt-in to specific, aggressive rules (similar to the nursery group), but we haven't implemented that yet, the main challenge being explaining and communicating that behavior to users.

**Comment by tjkuson** [comment]
> I think I'd prefer not to implement this rule for now. It requires a Pylint extension (`pylint.extensions.while_used`), so it's not enabled by default, and I suspect most users would find it to be too aggressive to be on-by-default.

That's reasonable. I can see how it would be unhelpful for many users!
 
> I'd be open to implementing it if we had the ability for users to selectively opt-in to specific, aggressive rules (similar to the nursery group), but we haven't implemented that yet, the main challenge being explaining and communicating that behavior to users.

Unless you think it's unwise, I will keep the PR open. If a rule opt-in or other feature were to be added, I think this rule would be a valuable addition for users like myself who prefer to avoid `while` loops (if not demonstrated by the fact it's a built-in Pylint extension).

**Comment by tjkuson** [comment]
> the main challenge being explaining and communicating that behavior to users.

Are you open to the idea of creating a namespace for esoteric rules? For example, it would be confusing if selecting “PL” didn't also select “PLW0149”. But, if the opt-in rules were under, for example, namespace “X” (the universal character for spooky), then it wouldn't be too surprising that the behaviour for “XPLW0149” were different.

**Comment by charliermarsh** [comment]
> Unless you think it's unwise, I will keep the PR open.

I'm going to close this, only because I like to try and keep the PR queue actionable. (I don't always do a great job of this, but I do _try_.) If we want to revive this work, we'll always have the history here as a closed PR -- but by closing, we ensure that we're not keeping around too many PRs that aren't being actively worked on.

> Are you open to the idea of creating a namespace for esoteric rules? 

I'm open to the idea, but I don't know that I'd want to layer it on to the existing categorizations (e.g., `XPLW0149`). I'd probably rather have a pedantic category that isn't included in `ALL` or similar.

**Pull request closed without merging**